### ISSUE TITLE
Consume and read loop causes memory leak

### ISSUE THREAD

**Issue by relu91** [eclipse-thingweb/node-wot#1062]
I discovered this memory leak a week ago, but only today I finally got what is the issue. Take this simple code:
```js
const { Servient } = require("@node-wot/core");
const { HttpClientFactory } = require("@node-wot/binding-http");
const { Helpers } = require("@node-wot/core");

let WoTHelpers;
let api;
async function main(){
    const servient = new Servient();
    servient.addClientFactory(new HttpClientFactory());
    api = await servient.start();
    WoTHelpers = new Helpers(servient);
    readloop();
}

async function readloop() {
    const td = await WoTHelpers.fetch("http://localhost:8080/test")
    const testThing = await api.consume(td);
    console.log("Consumed");

    const output = await testThing.readProperty("temperature")
    const value = await output.value();

    console.log(value);
    setTimeout(() => {
        main();
    }, 500);
}

main();
```

As you can see, we are simply polling a remote web thing and reading its `temperature` property. The loop seems pretty harmless but as time goes by the heap starts to inflate and after some days your service will crash for not having free available memory. 

Putting aside the reason why we are continuously fetching the consumed thing (more on that later), the root cause of this memory leak is `ajv` cache. Basically, every time we call the `value` function we call `ajv.compile(schema)` which, as a side effect, stores the result of the compilation inside an internal `Map` cache. The schema object is used as a key of this `Map` hence every time we fetch and consume the Thing Description we are creating a new schema object which for the map is a completely new key. This also triggers another compilation round which has an impact also in the CPU performance. 

One easy solution for our users is simply to always avoid fetch-consume loops (btw the same applies to simply consuming the same Thing Description in a loop because we are cloning it every time). However, in my use case, we were continuously fetching to have some small tolerance in transient services shutdowns ( e.g. if the remote thing updates the consumer service recovers automatically ) .

One solution could be to handle the caching of the Thing Description schemas ourselves (`ajv` provides a good API for managing schemas manually), but this would require some major efforts since, in the current architecture, the `InteractionOutput` has the responsibility of handling schema validation and it has a narrow view of the whole ConsumeThing. We would need to move the compilation of schemas responsibility up to the ConsumedThing object and pass it as a parameter to the  `InteractionOutput` object which only calls the validate function. 

Any thoughts?

**Comment by egekorkan** [comment]
I feel that the best approach would be to manage the compiled TD schema ourselves and not compile (or even look for keys) each time but the same might be more difficult for each interaction.

**Comment by danielpeintner** [comment]
Note: I am not very familiar with ajv but looking at our code there are essentially 4 packages where we call `ajv.compile`

* `binding-opcua`
* `cli`  ... `cli.ts` 
* `td-tools`  .... `thing-model-helpers.ts` 
* `core`  ... `helpers.ts` and `interaction-output.ts`

I think the problem arises from the latter in `interaction-output.ts` which is called very often, right? 
https://github.com/eclipse-thingweb/node-wot/blob/3a745126c7513b84e185c15e44b293c0cdda421d/packages/core/src/interaction-output.ts#L106


Currently ajc is initialized once/globally, on line 32 .. wouldn't it help to do it locally?
https://github.com/eclipse-thingweb/node-wot/blob/3a745126c7513b84e185c15e44b293c0cdda421d/packages/core/src/interaction-output.ts#L32

I mean every time on line 106 we could create an ajv instance and there should be no leak anymore... anyhow... it might slow down processing.. for sure.
Am I wrong or do I misunderstand the problem?

**Comment by egekorkan** [comment]
I think that initializing it each time would solve the leak issue but would decrease the performance. Another option (that I do not like either) would be to reinitialize ajv _once in a while_.

**Comment by danielpeintner** [comment]
> Another option (that I do not like either) would be to reinitialize ajv _once in a while_.

Yes, doing it once every 1_000 calls or so would not hurt performance and still keep the memory usage in a certain bound...
Let's wait for @relu91 and what he thinks... to me it seems to be a quick (but maybe dirty?) solution 🙃

**Comment by danielpeintner** [comment]
My proposal would be two-fold:
1. SOON re-initializing ajv every time (or every 1000 times or so, either way is fine since the performance penalty would be very minimal and the memory leak is somewhat limited)
2. LATER looking into this issue more closely
2.X *Personally I am not sure whether it is worth to to anything more.. causing too much complexity ... but I am happy to be taught better*

**Comment by relu91** [comment]
I dug a little bit more into this issue, seems we are not alone: https://github.com/ajv-validator/ajv/issues/1413. In the discussion, it seems that we can avoid re-initializing AJV every time, we could simply use `addUsedSchema: false`. However, caching is strongly recommended. We can go on fixing the memory leak but we should plan for the caching feature after this release.

**Comment by danielpeintner** [comment]
Since at the moment we seem to not have time to work on the caching feature I suggest to fix the solution now (see https://github.com/eclipse-thingweb/node-wot/pull/1266) and keep this issue to be _properly_ solved in the future.
